server/cri/v1/rpc_container_events.go

```diff
@@ -5,6 +5,13 @@ import (
 )
 
 func (s *service) GetContainerEvents(req *pb.GetEventsRequest, ces pb.RuntimeService_GetContainerEventsServer) error {
-	// Spoofing the GetContainerEvents method to return nil response until actual support code is added.
+
+	if s.server.Config().EventedPLEG {
+		for containerEvent := range s.server.ContainerEventsChan {
+			if err := ces.Send(&containerEvent); err != nil {
+				return err
+			}
+		}
+	}
```

Comment:
Is this really how this should work ? Maybe I misread the code, but to me it looks like that you are
  - using `s.server.ContainerEventsChan` to buffer at most a 1000 events at a time
  - you send all pending/'buffered' events from the channel to the stream when you get a `GetContainerEvents()`
  - you never send any subsequent events to the stream (even if the client sticks around)

I might be totally wrong here, but my impression of the intention of this new CRI API was that when a client makes a `GetContainerEvents()` call, it should keep receiving container events over the stream until the client itself goes away (or until the server decides to stop sending events and close the stream). And that there might be several concurrent clients listening for container events...

Reply:
Again IIUC, the exhibited behavior of this implementation will be that when a client does a `GetContainerEvents()`, it will get all the queued events from the `ContainerEventsChan` and then receive nothing ever after. Also, if there has been more than a 1000 events emitted/generated by that time, the client will get the oldest 1000 with anything newer being already lost.

So shouldn't this instead work somehow so that:
- in `GetContainerEvents()` you simply store the new client stream along with any other existing ones
- you set up an 'event pump' goroutine which sits in loop reading events from `ContainerEventsChan` and multiplexes them to all client streams
- in the event pump you close/remove any client stream on which sending an event fails (client gone, or too slow to read)

Reply:
+1 to a multiplexed approach

Reply:
@klihub, @haircommander  Thanks for the suggestions, let me check this out and get back.